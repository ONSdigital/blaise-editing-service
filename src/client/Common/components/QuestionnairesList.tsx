import { ReactElement, SetStateAction, useState } from 'react';
import { ONSSelect } from 'blaise-design-system-react-components';
import { User } from 'blaise-api-node-client';
import { QuestionnaireDetails } from '../../../common/interfaces/surveyInterface';
import Option from '../../Interfaces/controlsInterface';
import UserRole from '../enums/UserRole';
import SupervisorsQuestionnaireDetails from '../../Supervisor/Components/SupervisorsQuestionnaireDetails';
import EditorsQuestionnaireDetails from '../../Editor/Components/EditorsQuestionnaireDetails';
import ErrorPanel from './ErrorPanel';

interface QuestionnairesListProps {
  questionnaires: QuestionnaireDetails[];
  user: User
}

function RenderQuestionnaireDetails(user:User, questionnaire:QuestionnaireDetails) {
  const { role, name } = user;

  if (role === UserRole.SVT_Supervisor) {
    return <SupervisorsQuestionnaireDetails questionnaire={questionnaire} supervisorRole={UserRole.SVT_Supervisor} editorRole={UserRole.SVT_Editor} />;
  }

  if (role === UserRole.SVT_Editor) {
    return <EditorsQuestionnaireDetails questionnaire={questionnaire} username={name} editorRole={UserRole.SVT_Editor} />;
  }

  return <ErrorPanel message={`User role ${role} not recognised`} />;
}

function getquestionnaireOptions(questionnaires: QuestionnaireDetails[]): Option[] {
  const options: Option[] = [];

  questionnaires.forEach((questionnaire) => {
    options.push({
      label: `${questionnaire.questionnaireName} (${questionnaire.fieldPeriod})`,
      value: questionnaire.questionnaireName,
    });
  });

  return options;
}

function getDefaultQuestionnaire(questionnaires: QuestionnaireDetails[]) : QuestionnaireDetails {
  const defaultQuestionnaire = questionnaires[0];
  if (defaultQuestionnaire === undefined) {
    throw Error('Questionnaires');
  }

  return defaultQuestionnaire;
}

function getQuestionnaire(questionnaires: QuestionnaireDetails[], questionnaireName?: string) : QuestionnaireDetails {
  return questionnaires.find((q) => q.questionnaireName === questionnaireName) ?? getDefaultQuestionnaire(questionnaires);
}

function getDefaultQuestionnaireOptionValue(questionnaires: QuestionnaireDetails[]): string {
  const savedQuestionnaireValue = localStorage.getItem('savedQuestionnaireOption') ?? undefined;
  const defaultQuestionnaire = getQuestionnaire(questionnaires, savedQuestionnaireValue);

  return defaultQuestionnaire.questionnaireName;
}

export default function QuestionnairesList({ questionnaires, user }: QuestionnairesListProps): ReactElement {
  const [questionnaireValue, setQuestionnaireValue] = useState(getDefaultQuestionnaireOptionValue(questionnaires));

  const handleQuestionnaireChange = (e: { target: { value: SetStateAction<string>; }; }) => {
    setQuestionnaireValue(e.target.value);
    localStorage.setItem('savedQuestionnaireOption', e.target.value.toString());
  };

  return (
    <>
      <ONSSelect
        defaultValue={questionnaireValue}
        id="select-questionnaire"
        label="Select questionnaire"
        options={getquestionnaireOptions(questionnaires)}
        value=""
        onChange={handleQuestionnaireChange}
        testId="select-questionnaire"
      />
      {
        RenderQuestionnaireDetails(user, getQuestionnaire(questionnaires, questionnaireValue))
      }
    </>
  );
}
