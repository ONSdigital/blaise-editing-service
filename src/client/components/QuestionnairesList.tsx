import { Questionnaire } from 'blaise-api-node-client';
import { ONSPanel, ONSTable } from 'blaise-design-system-react-components';
import { Link } from 'react-router-dom';

interface QuestionnairesListProps {
  questionnaires: Questionnaire[];
}

export default function QuestionnairesList({ questionnaires }: QuestionnairesListProps) {
  if (questionnaires.length === 0) {
    return (
      <ONSPanel spacious status="info">There are no questionnaires available</ONSPanel>
    );
  }
  return (
    <ONSTable
      tableID="questionnaire-table"
      columns={[
        'Questionnaire name',
        'Server park',
        'Number of cases',
      ]}
    >
      <>
        {questionnaires.map((questionnaire) => (
          <tr key={questionnaire.name} className="ons-table__row" data-testid="questionnaire-table-row">
            <td className="ons-table__cell"><Link to={`/questionnaires/${questionnaire.name}/cases`}>{questionnaire.name}</Link></td>
            <td className="ons-table__cell">{questionnaire.serverParkName}</td>
            <td className="ons-table__cell">{questionnaire.dataRecordCount}</td>
          </tr>
        ))}
      </>
    </ONSTable>
  );
}
