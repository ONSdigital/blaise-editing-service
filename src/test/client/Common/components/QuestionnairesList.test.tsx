import { render, RenderResult } from '@testing-library/react';
import userMockObject from '../../../server/mockObjects/userMockObject';
import QuestionnairesList from '../../../../client/Common/components/QuestionnairesList';
import { frsQuestionnaireDetails1MockObject } from '../../../server/mockObjects/questionnaireListMockObject';

let view:RenderResult;

describe('Renders the error screen if the users role is not in the accepted list', () => {
  it('should render the editor page correctly when given an invalid role', async () => {
    // arrange
    const user = userMockObject;
    const questionnaire = frsQuestionnaireDetails1MockObject;
    user.role = 'invalid_Role';

    // act
    view = render(<QuestionnairesList questionnaires={[questionnaire]} user={user} />);

    // assert
    expect(view).toMatchSnapshot(
      'EditorPageInvalidRole',
    );
  });

  it('Should display the expected error on the screen when given an invalid role', async () => {
    // arrange
    const user = userMockObject;
    const questionnaire = frsQuestionnaireDetails1MockObject;
    user.role = 'invalid_Role';

    // act
    view = render(<QuestionnairesList questionnaires={[questionnaire]} user={user} />);

    // assert
    const errorDetails = view.getByTestId('ErrorMessage');
    expect(errorDetails).toHaveTextContent('User role invalid_Role not recognised');
  });
});
