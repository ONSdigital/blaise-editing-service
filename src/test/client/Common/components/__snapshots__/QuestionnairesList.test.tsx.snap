// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders the error screen if the users role is not in the accepted list should render the editor page correctly when given an invalid role 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div>
        <label
          class="ons-label"
          for="select-questionnaire"
        >
          Select questionnaire
           
        </label>
        <select
          class="ons-input "
          data-testid="select-questionnaire"
          id="select-questionnaire"
          name="select"
        >
          <option
            data-testid="select-select-questionnaire"
            disabled=""
            value=""
          >
            Select an option
          </option>
          <option
            data-testid="option-select-questionnaire-FRS2408B_EDIT"
            selected=""
            value="FRS2408B_EDIT"
          >
            FRS2408B (August 2024)
          </option>
        </select>
      </div>
      <div
        class="ons-panel ons-panel--error ons-panel--no-title  ons-u-mt-m"
      >
        <div
          class="ons-panel__body "
        >
          <h4>
            Something went wrong
          </h4>
          <p
            data-testid="ErrorMessage"
          >
            User role invalid_Role not recognised
          </p>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div>
      <label
        class="ons-label"
        for="select-questionnaire"
      >
        Select questionnaire
         
      </label>
      <select
        class="ons-input "
        data-testid="select-questionnaire"
        id="select-questionnaire"
        name="select"
      >
        <option
          data-testid="select-select-questionnaire"
          disabled=""
          value=""
        >
          Select an option
        </option>
        <option
          data-testid="option-select-questionnaire-FRS2408B_EDIT"
          selected=""
          value="FRS2408B_EDIT"
        >
          FRS2408B (August 2024)
        </option>
      </select>
    </div>
    <div
      class="ons-panel ons-panel--error ons-panel--no-title  ons-u-mt-m"
    >
      <div
        class="ons-panel__body "
      >
        <h4>
          Something went wrong
        </h4>
        <p
          data-testid="ErrorMessage"
        >
          User role invalid_Role not recognised
        </p>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
